lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -137,9 +147,46 @@ static int createAsyncCopy(scf::ForOp &forOp, tt::LoadOp loadOp, Value alloc,
       mask = convertBlockLayout(mask);
     if (other)
       other = convertBlockLayout(other);
+  } else if (loadToInfo[loadOp].isMMAv3Registers) {
+    // If the following are true...
+    //   1) Operand A is for WGMMA and is to be loaded in registers
+    //   2) We upcast operand A in registers before the WGMMA
+    //      (downcasting is not yet supported)
+    //
+    // ...then the SharedEncoding vec will be less than BlockedEncoding's
+    // sizePerThread for k-dim. E.g. if shared vec is 8 and sizePerThread
+    // for k is 16, then AsyncCopyGlobalToLocal will generate two 8B-LDGSTS's
+    // for each contiguous 16B global data owned by each thread. This breaks
+    // coalescing (i.e. results 2x the minimum required transactions)
+    //
+    // The fix is to clip the BlockedEnc's sizePerThread using SharedEnc's vec.
+    auto tensorTy = cast<RankedTensorType>(src.getType());
+    auto blockEnc = cast<ttg::BlockedEncodingAttr>(tensorTy.getEncoding());
+    auto sharedEnc = cast<ttg::SharedEncodingAttr>(allocTy.getEncoding());
+    auto sharedVec = sharedEnc.getVec();
+
+    SmallVector<unsigned> newSizePerThread;
+    llvm::transform(blockEnc.getSizePerThread(),
+                    std::back_inserter(newSizePerThread),
+                    [&](auto size) { return std::min(size, sharedVec); });
+
+    if (newSizePerThread != blockEnc.getSizePerThread()) {
+      auto mod = loadOp->getParentOfType<ModuleOp>();
+      int numWarps = ttg::TritonGPUDialect::getNumWarps(mod);
+      int threadsPerWarp = ttg::TritonGPUDialect::getThreadsPerWarp(mod);
+      auto newBlockEnc = ttg::BlockedEncodingAttr::get(
+          loadOp.getContext(), tensorTy.getShape(), newSizePerThread,
+          blockEnc.getOrder(), numWarps, threadsPerWarp,
+          blockEnc.getCTALayout());
+
+      src = convertBlockLayout(src, newBlockEnc);
+      if (mask)
+        mask = convertBlockLayout(mask, newBlockEnc);
+      if (other)
+        other = convertBlockLayout(other, newBlockEnc);
```

Comment:
that looks like an overfit on a special cases, there are many cases where this would not be beneficial. This is probably some optimization we want to do later in the flow with a more robust heuristic. Can this be done in a separate PR and as a separate optimization?

Reply:
Sure, I've removed this for now. I did tentative measurements for a few shapes without this above optimization:
(in microseconds)
| Shape | Main | Current PR w/ this optimizations | Current PR w/o this optimization 
| ---------| ------------- | ------------- | ---------------|
| 8, 8192, 8192 |  31|  28 | 29|
| 2048, 2048, 2048 |  340  | 290  | 332 |
| 4096, 4096, 4096 | 2626  | 2459  | 2625 |

So while this optimization is helpful, without it, we still see _slight_ gains over baseline.

I can start working on a separate PR, although I'm trying to think how I can make this logic more general. If shared vec < blocked sizePerThread, I'm not sure if there's a way to fix coalescing other than reducing sizePerThread for the blocked encoding. Do you have specific cases in mind where this would be harmful?

Reply:
yes I think reducing sizePerThread is fine, I think what we want to avoid is, doing to much in the pipeliner and making assumptions on how the IR is just based on the fact that the loads feed into mma through register.

Reply:
yep, makes sense. in terms of implementation, should I create a separate pass after the pipeliner? (I'm not sure it makes sense to be put into any of the existing passes)

Reply:
a friendly ping for advice on where to best place this optimization

Reply:
sorry for the delay. 
This problem already exists if we don't pipeline and don't use async.cp right? If that's the case I think the best place would be to do it as an improvement of the coalescing pass

Reply:
I've only observed this with async copy actually.

In the int8 example where shared-vec == 8 and blocked-sizePerThread == 16, if pipelining is disabled, then loading is done with `LDG.128` (128 = 16 * 8 bits), the data is cast to f16, and storing is done with `STS.128` (128 = 8 * 16bits).

If pipelining is enabled, the `LDG` and `STS` are coupled. Meaning we can't load 16 int8's in one 128-bit instruction. Two `LDGSTS.64` are generated for the contiguous 128 bits, which breaks coalescing.

Hence, I was thinking this might make sense as a new pass somewhere after the pipeliner?

Reply:
I see, I agree it should go in a pass after pipelining then. I can't think of a good place for this right now so I think making a new pass makes sense and we can move it around over time.